### PR TITLE
Speed up docker build times

### PR DESCRIPTION
**Pull request by karlkfi**
- Split the docker images into their own Dockerfiles, to avoid re-building the base image and every other image for each component.
- Build the base "bins" image locally, but don't push it.
- This should cut total image build time in half.

### CODE REVIEW

**Comment by sdowell** [comment]
I did some quick timings locally, clearing the image cache before each run to test with a clean initial state:
```
(HEAD detached at upstream/main) $ time make build-images
go mod tidy -compat=1.17
+++ Building the Reconciler image: gcr.io/sdowell-cs-dev/sdowell/reconciler:v1.14.0-rc.2-42-g498a6da6f
sha256:9351b523efe0a74f0220b1fab7b4805b4166e84a2290a4390e5e6f40508d4374
+++ Building the Reconciler Manager image: gcr.io/sdowell-cs-dev/sdowell/reconciler-manager:v1.14.0-rc.2-42-g498a6da6f
sha256:8f781d041e3ff83a97a2418a29a86286fbe522f66597a69c21d6fecb59f69827
+++ Building the Admission Webhook image: gcr.io/sdowell-cs-dev/sdowell/admission-webhook:v1.14.0-rc.2-42-g498a6da6f
sha256:0dc77ceeb605fa02fd93fe881a8f496189afa1a7364f966c24c82e8a194335f1
+++ Building the Hydration Controller image: gcr.io/sdowell-cs-dev/sdowell/hydration-controller:v1.14.0-rc.2-42-g498a6da6f
sha256:852899b92be1b2a8623e0c2b4845e67fe33ff664cc1e68df344bf866a5428e5d
+++ Building the Hydration Controller image with shell: gcr.io/sdowell-cs-dev/sdowell/hydration-controller-with-shell:v1.14.0-rc.2-42-g498a6da6f
sha256:6c73d0f9ec4bc978bd1cc5acccf3277557ad27f754816be344c12dcf7068c607
+++ Building the OCI-sync image: gcr.io/sdowell-cs-dev/sdowell/oci-sync:v1.14.0-rc.2-42-g498a6da6f
sha256:387cf371c64b7a45d12007b7417d12811838b9d1a379f6ecfa69833bdc3d340e
+++ Building the Helm-sync image: gcr.io/sdowell-cs-dev/sdowell/helm-sync:v1.14.0-rc.2-42-g498a6da6f
sha256:a5bf803cb0808b8195a058a67e6c2f79c7b72ff1fb16cbe24f6a633ff5aed679

real    7m15.059s
user    0m49.186s
sys     0m35.499s
```
vs
```
(HEAD detached at karl/karl-build-speed) $ time make build-images
go mod tidy -compat=1.17
+++ Building the Bins image: local/local/kpt-config-sync-bins:v1.14.0-rc.2-43-g7d2ded919
sha256:19aa211931ef6b926281fc75d1bbf40853762b0b789d0882e88cef37cded15a8

+++ Building the Reconciler image: gcr.io/sdowell-cs-dev/sdowell/reconciler:v1.14.0-rc.2-43-g7d2ded919
sha256:a642b9e186966b89c9b519fc1ef61ce64209a1f431e0450132cde888205f8195

+++ Building the Reconciler Manager image: gcr.io/sdowell-cs-dev/sdowell/reconciler-manager:v1.14.0-rc.2-43-g7d2ded919
sha256:6989a731ee9c506e54e547c2e717991c404463cf314b0f6def77602b8767a84c

+++ Building the Admission Webhook image: gcr.io/sdowell-cs-dev/sdowell/admission-webhook:v1.14.0-rc.2-43-g7d2ded919
sha256:d6490695dc4a3210a065bfb7b09c54bb92ce21b62f246d3ab743a2a99370699b

+++ Building the Hydration Controller image: gcr.io/sdowell-cs-dev/sdowell/hydration-controller:v1.14.0-rc.2-43-g7d2ded919
sha256:d8e9cf9cb29a21e311c3bf132f5a3db2d0839bc9a285ad15358708120d3bfe9f

+++ Building the Hydration Controller image with shell: gcr.io/sdowell-cs-dev/sdowell/hydration-controller-with-shell:v1.14.0-rc.2-43-g7d2ded919
sha256:3c2585080daaa9b12fa6a108ed8052d58716f065c4420050b8051e03f6bf9cdf

+++ Building the OCI-sync image: gcr.io/sdowell-cs-dev/sdowell/oci-sync:v1.14.0-rc.2-43-g7d2ded919
sha256:37ff4963f56ca20d35d4435fce4f8fb8e16cdb544dfde56ef6fafee0e6fcf1c2

+++ Building the Helm-sync image: gcr.io/sdowell-cs-dev/sdowell/helm-sync:v1.14.0-rc.2-43-g7d2ded919
sha256:6c1deb61ad06f381fa03e8a30f71bc304b9aa5d51a60e10cdaa5571e46acb8ba

real    7m9.053s
user    1m1.506s
sys     0m44.789s
```
Looks like not much difference

**Comment by sdowell** [comment]
Timed using `docker buildx build` locally:
```
$ time make build-images
go mod tidy -compat=1.17
+++ Building the reconciler image: gcr.io/sdowell-cs-dev/sdowell/reconciler:v1.14.0-rc.2-43-g5b5a0ef43-dirty
sha256:0e88f70af469aaafc38843acc1b75984d745cae9b6d9f2e88c630e8b9dff6f84
+++ Building the reconciler-manager image: gcr.io/sdowell-cs-dev/sdowell/reconciler-manager:v1.14.0-rc.2-43-g5b5a0ef43-dirty
sha256:06e0830bc3206581cec41de8475ab00581bd94175b70f6e5549e21e3bdd5e6f4
+++ Building the admission-webhook image: gcr.io/sdowell-cs-dev/sdowell/admission-webhook:v1.14.0-rc.2-43-g5b5a0ef43-dirty
sha256:c4e9d1462bb91fbde4dc4b57cd3a9665435372e66624fa16dad4d0142f5951c0
+++ Building the hydration-controller image: gcr.io/sdowell-cs-dev/sdowell/hydration-controller:v1.14.0-rc.2-43-g5b5a0ef43-dirty
sha256:e563063c00a601c790cdc9503ee9527174234522a89aad57fe36cc24165bb862
+++ Building the hydration-controller-with-shell image: gcr.io/sdowell-cs-dev/sdowell/hydration-controller-with-shell:v1.14.0-rc.2-43-g5b5a0ef43-dirty
sha256:819322f1865f6be388216ad4e31d50305804bf80a93fa94f1e9af651361be9d7
+++ Building the oci-sync image: gcr.io/sdowell-cs-dev/sdowell/oci-sync:v1.14.0-rc.2-43-g5b5a0ef43-dirty
sha256:9a63baf14ae38df7db37ec9db6ee794113bc286361935e0f3729a14460a16d12
+++ Building the helm-sync image: gcr.io/sdowell-cs-dev/sdowell/helm-sync:v1.14.0-rc.2-43-g5b5a0ef43-dirty
sha256:c89c2d6a5a3796dae8c98e45a5d1eab1e57c075ab2505441a1afe4c15e9b939b

real    2m4.342s
user    0m30.517s
sys     0m22.526s
```

**Comment by karlkfi** [comment]
@sdowell can you add a time call before each build and run those again? My times were 11m vs 5m. I’m wondering where the time is being spent.

**Comment by sdowell** [comment]
@karlkfi Added the following for each `docker build` call for finer grained timing metrics:
```sh
@time --format="took %E" docker build ...
```
And pruning the image/builder cache before each time:
```bash
(HEAD detached at karl/karl-build-speed) $ time make build-images
go mod tidy -compat=1.17
+++ Building the Bins image: local/local/kpt-config-sync-bins:v1.14.0-rc.2-43-g7d2ded919-dirty
sha256:3ddc5e8c002f343bbe4db293231b384dc2ef60fd00b6d32224ef3e3e5b1d5f07
took 2:55.30

+++ Building the Reconciler image: gcr.io/sdowell-cs-dev/sdowell/reconciler:v1.14.0-rc.2-43-g7d2ded919-dirty
sha256:83fa96d4adf72a5daab6a791b9788bd3346127d0966a0283bb01d9e89d376791
took 0:32.49

+++ Building the Reconciler Manager image: gcr.io/sdowell-cs-dev/sdowell/reconciler-manager:v1.14.0-rc.2-43-g7d2ded919-dirty
sha256:f194f653363145b379db7c00809364216694ffe2d287522202408340249171c3
took 0:32.52

+++ Building the Admission Webhook image: gcr.io/sdowell-cs-dev/sdowell/admission-webhook:v1.14.0-rc.2-43-g7d2ded919-dirty
sha256:b15b38841ba0fb577be018c9d33056f85ab73103c1800900243321a236bda6bb
took 0:32.30

+++ Building the Hydration Controller image: gcr.io/sdowell-cs-dev/sdowell/hydration-controller:v1.14.0-rc.2-43-g7d2ded919-dirty
sha256:e6e7a0d6d2878f8ea358ed5169cc0e97d09ae303cbfe4f29fe1645c347665a93
took 0:34.21

+++ Building the Hydration Controller image with shell: gcr.io/sdowell-cs-dev/sdowell/hydration-controller-with-shell:v1.14.0-rc.2-43-g7d2ded919-dirty
sha256:73850561babdf4dc918394164de3d0f3a3740ba4e0c2ce43237476853b083af5
took 0:52.58

+++ Building the OCI-sync image: gcr.io/sdowell-cs-dev/sdowell/oci-sync:v1.14.0-rc.2-43-g7d2ded919-dirty
sha256:c268fa5b8cffe8d48c2995c4c640ecdf470644cf87d18f002679ee8ab739fcbd
took 0:31.00

+++ Building the Helm-sync image: gcr.io/sdowell-cs-dev/sdowell/helm-sync:v1.14.0-rc.2-43-g7d2ded919-dirty
sha256:3408e76b608b52854137c85c679e0f24423b4fb725536d4bbda9e948998fe8a6
took 0:30.76

real    7m2.516s
user    0m58.448s
sys     0m41.050s
```
vs
```bash
(HEAD detached at upstream/main) $ time make build-images
go mod tidy -compat=1.17
+++ Building the Reconciler image: gcr.io/sdowell-cs-dev/sdowell/reconciler:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:a40c18435e16bca25b1e5f92ba0672a208c8f70528f36acd05d40a0d0b4882d9
took 3:30.10
+++ Building the Reconciler Manager image: gcr.io/sdowell-cs-dev/sdowell/reconciler-manager:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:054397622a27247b46d3622f595f8fa68098e1c5b93eb955294d2bc2ce17cbe8
took 0:41.34
+++ Building the Admission Webhook image: gcr.io/sdowell-cs-dev/sdowell/admission-webhook:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:490a5d4db14fe04e471959ee9c3e1bc6fb6a96097f8b2b85f95ddfeba9275bcb
took 0:41.57
+++ Building the Hydration Controller image: gcr.io/sdowell-cs-dev/sdowell/hydration-controller:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:a3219a274983c2da7afe82963c296ce4bcd7c47b0f9f8b23addee423a09c1b79
took 0:38.97
+++ Building the Hydration Controller image with shell: gcr.io/sdowell-cs-dev/sdowell/hydration-controller-with-shell:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:6869020652403d2d85ec40c8fa1cff738924e2310318b96ccbc7e3db57ff9a91
took 0:39.87
+++ Building the OCI-sync image: gcr.io/sdowell-cs-dev/sdowell/oci-sync:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:7db6b0417dbf82f5251e0c3aad702cb9f64996b8214ceae302217c3a403f1be8
took 0:39.11
+++ Building the Helm-sync image: gcr.io/sdowell-cs-dev/sdowell/helm-sync:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:34ee5f224d108577b143f22c94526e12207a20433ddfe5681accf7f81bc96eb1
took 0:38.74

real    7m30.943s
user    0m52.173s
sys     0m37.087s
```
TL;DR looks like it's already using the cached bins layer from the multi-stage build after the first build (`reconciler`).

**Comment by sdowell** [comment]
Same experiment using `docker buildx build`:
```bash
(HEAD detached at upstream/main) $ time make build-images
go mod tidy -compat=1.17
+++ Building the Reconciler image: gcr.io/sdowell-cs-dev/sdowell/reconciler:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:0dd8620e16eca556337c2698feff8b414e96377ee15fee3136d64565555e0e22
took 1:46.04
+++ Building the Reconciler Manager image: gcr.io/sdowell-cs-dev/sdowell/reconciler-manager:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:1522012eaa4ac41ea39a37efb65efce7f744975eeb98a0410dbe6759b5afb047
took 0:03.93
+++ Building the Admission Webhook image: gcr.io/sdowell-cs-dev/sdowell/admission-webhook:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:1adb127cb15baace5cded6bad9b676d882c5ac816d1004bd912ae108cc1bc861
took 0:03.86
+++ Building the Hydration Controller image: gcr.io/sdowell-cs-dev/sdowell/hydration-controller:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:175b915e08420f285a8ffc808ed367e4e47e16163e12aed9992739bae3799124
took 0:04.81
+++ Building the Hydration Controller image with shell: gcr.io/sdowell-cs-dev/sdowell/hydration-controller-with-shell:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:580b37fef3a2ec995b6f957bee191bf227a50cafa4ea89df0211aab0a6792ee0
took 0:18.77
+++ Building the OCI-sync image: gcr.io/sdowell-cs-dev/sdowell/oci-sync:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:43e8f29c4072ccdd3c563bb58020acd4f690891e32f756e4b2d67a941baa5d6d
took 0:04.06
+++ Building the Helm-sync image: gcr.io/sdowell-cs-dev/sdowell/helm-sync:v1.14.0-rc.2-42-g498a6da6f-dirty
sha256:9d80601e39eda2666a2cb4b91be943c6a099449c5cb57ccf8eba7c541048fb28
took 0:04.12

real    2m26.996s
user    0m35.332s
sys     0m28.289s
```

**Comment by karlkfi** [comment]
Obsolete. Replaced by https://github.com/GoogleContainerTools/kpt-config-sync/pull/322
